metatensor-core/include/metatensor/tensor.hpp

```diff
@@ -13,6 +16,96 @@
 
 namespace metatensor {
 
+namespace details {
+
+/// An iterator for TensorMap info key/values.
+///
+/// This class is not intended for direct usage, only through the
+/// `TensorMap::info()` function. The iterator yields
+/// `std::pair<std::string_view, std::string_view>`.
+///
+/// This class and any element obtained by iteration are only valid as long as
+/// the underlying `TensorMap` is alive, and no changes are made to the info
+/// map. If you need to keep the data longer, you should make a copy of the key
+/// and value strings.
+///
+/// ```cpp
+/// for (auto [key, value]: tensor_map.info()) {
+///     // do something with key and value
+/// }
+/// ```
+class TensorMapInfo {
+public:
+    /// @private Create an iterator for the given `tensor`
+    explicit TensorMapInfo(mts_tensormap_t* tensor) : tensor_(tensor) {
+        details::check_status(mts_tensormap_info_keys(
+            tensor_,
+            &keys_,
+            &count_
+        ));
+    }
+
+    /// @private
+    class iterator {
+    public:
+        // iterator traits
+        using difference_type = std::ptrdiff_t;
+        using value_type = std::pair<std::string_view, std::string_view>;
+        using pointer = const value_type*;
+        using reference = value_type;
+        using iterator_category = std::forward_iterator_tag;
```

Comment:
```suggestion
        using iterator_category = std::input_iterator_tag;
```

because of the return (creating a new value on `operator*`).

xref: https://en.cppreference.com/w/cpp/iterator/input_iterator.html